clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java

```diff
@@ -0,0 +1,193 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Abstract assignor implementation which uses the following metadata format:
+ *
+ * ProtocolMetadata => Topics TopicPattern MetadataHash
+ *   Topics         => [String]
+ *   TopicPattern   => String
+ *   MetadataHash   => Bytes
+ *
+ * This class handles metadata consistency checking between group members. Custom assignors will
+ * usually extend this class unless they need to provide their own metadata format.
+ */
+public abstract class AbstractPartitionAssignor implements PartitionAssignor<AbstractPartitionAssignor.ConsumerMetadata> {
+    private static final String TOPIC_KEY_NAME = "topics";
+    private static final String TOPIC_PATTERN_KEY_NAME = "topic_pattern";
+    private static final String METADATA_HASH_KEY_NAME = "metadata_hash";
+
+    public static final Schema CONSUMER_METADATA_SCHEMA = new Schema(
+            new Field(TOPIC_KEY_NAME, new ArrayOf(Type.STRING)),
+            new Field(TOPIC_PATTERN_KEY_NAME, Type.STRING),
+            new Field(METADATA_HASH_KEY_NAME, Type.BYTES));
+
+    private static final Type CONSUMER_METADATA_TYPE = new ConsumerMetadataType();
+
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public abstract List<TopicPartition> assign(String consumerId,
+                                                Map<String, Integer> partitionsPerTopic,
+                                                Map<String, ConsumerMetadata> consumerMetadata);
+
+    public Type schema() {
+        return CONSUMER_METADATA_TYPE;
+    }
+
+    @Override
+    public AssignmentResult assign(String consumerId,
+                                   Map<String, ConsumerMetadata> consumers,
+                                   MetadataSnapshot metadataSnapshot) {
+        byte[] topicMetadataHash = metadataSnapshot.hash();
+        Set<String> allSubscribedTopics = new HashSet<>();
+        boolean consistentMetadata = true;
+
+        for (ConsumerMetadata metadata : consumers.values()) {
+            byte[] memberTopicMetadataHash = Utils.toArray(metadata.topicMetadataHash);
+            consistentMetadata = consistentMetadata && Arrays.equals(topicMetadataHash, memberTopicMetadataHash);
+            allSubscribedTopics.addAll(metadata.topics);
+        }
+
+        if (!consistentMetadata) {
+            log.debug("Assignment failed due to inconsistent metadata among consumers");
+            return AssignmentResult.inconsistentMetadata(allSubscribedTopics);
+        }
+
+        Map<String, Integer> partitionsPerTopic = new HashMap<>();
+        for (String topic : allSubscribedTopics) {
+            Integer numPartitions = metadataSnapshot.topicMetadata().get(topic);
+            if (numPartitions != null)
+                partitionsPerTopic.put(topic, numPartitions);
+            else
+                log.debug("Skipping assignment for topic {} since no metadata is available", topic);
```

Comment:
Ok, that makes sense in that it allows consumers to make progress even if a topic needs to be created. If auto topic creation is not enabled, then this might be the only way to make progress if one of the topics doesn't yet exist. But doesn't this also mean that if you have auto topic creation enabled and the consumer group starts up before any producers, then it might have to wait an entire metadata refresh timeout before any data is consumed from that topic, even if producers start sending data to it immediately?

I think this is fine (and it's a one-time delay, 5 or 10 minutes by default IIRC), just want to make sure we understand the implications of doing this.